File: backend/core/src/modules/backlinks/api/platform-request-context.ts
import { createHmac, timingSafeEqual } from "node:crypto";

import type {
  FastifyInstance,
  FastifyReply,
  FastifyRequest,
} from "fastify";
import { z } from "zod";

import { createActorContext, type ActorContext } from "../domain/context/index.js";
import {
  BacklinkError,
  backlinkErrorCodes,
} from "../domain/errors/backlink-error.js";
import {
  backlinkProblemContentType,
  toBacklinkProblemDetails,
} from "./problem-details.js";

export const platformContextHeader = "x-growthos-platform-context";
export const platformContextSignatureHeader =
  "x-growthos-platform-context-signature";
export const platformContextVersion = "PlatformRequestContext.v1";
const platformContextIssuer = "growthos-platform-gateway";
const platformContextAudience = "growthos-backlinks-core";
const maximumLifetimeMs = 60_000;
const allowedClockSkewMs = 5_000;

const nonBlankIdentifier = z
  .string()
  .min(1)
  .max(200)
  .refine((value) => value === value.trim(), "Must not have surrounding whitespace");
const uniqueIdentifiers = (maximum: number) =>
  z
    .array(nonBlankIdentifier)
    .min(1)
    .max(maximum)
    .refine((values) => new Set(values).size === values.length, "Must be unique");
const permission = z
  .string()
  .min(3)
  .max(200)
  .regex(/^[a-z][a-z0-9_.-]*:[a-z][a-z0-9_.-]*$/);

export const platformRequestContextSchema = z
  .object({
    version: z.literal(platformContextVersion),
    issuer: z.literal(platformContextIssuer),
    audience: z.literal(platformContextAudience),
    issuedAt: z.string().datetime(),
    expiresAt: z.string().datetime(),
    correlationId: nonBlankIdentifier,
    actor: z
      .object({
        userId: nonBlankIdentifier,
        sessionId: nonBlankIdentifier,
        roles: uniqueIdentifiers(32),
      })
      .strict(),
    tenant: z
      .object({
        organizationId: nonBlankIdentifier,
        workspaceId: nonBlankIdentifier,
      })
      .strict(),
    project: z
      .object({
        websiteProjectId: nonBlankIdentifier,
        websiteProjectKey: nonBlankIdentifier,
      })
      .strict(),
    permissions: z
      .array(permission)
      .min(1)
      .max(128)
      .refine((values) => new Set(values).size === values.length, "Must be unique"),
  })
  .strict();

export type PlatformRequestContextV1 = Readonly<{
  version: typeof platformContextVersion;
  issuer: typeof platformContextIssuer;
  audience: typeof platformContextAudience;
  issuedAt: string;
  expiresAt: string;
  correlationId: string;
  actor: Readonly<{
    userId: string;
    sessionId: string;
    roles: readonly string[];
  }>;
  tenant: Readonly<{
    organizationId: string;
    workspaceId: string;
  }>;
  project: Readonly<{
    websiteProjectId: string;
    websiteProjectKey: string;
  }>;
  permissions: readonly string[];
}>;

declare module "fastify" {
  interface FastifyRequest {
    actor: ActorContext;
    platformContext: PlatformRequestContextV1;
  }
}

type ConsumerOptions = Readonly<{
  signingKey: string | Buffer;
  now?: () => Date;
}>;

function authenticationFailure(): BacklinkError {
  return new BacklinkError({
    code: backlinkErrorCodes.authenticationRequired,
    message: "A valid internal platform context is required.",
  });
}

function headerValue(
  request: FastifyRequest,
  header: string,
): string {
  const value = request.headers[header];
  if (typeof value !== "string" || value.length === 0) {
    throw authenticationFailure();
  }
  return value;
}

function decodeBase64Url(value: string): Buffer {
  if (!/^[A-Za-z0-9_-]+$/.test(value)) {
    throw authenticationFailure();
  }
  const decoded = Buffer.from(value, "base64url");
  if (decoded.toString("base64url") !== value) {
    throw authenticationFailure();
  }
  return decoded;
}

function verifyIntegrity(
  payload: string,
  signatureHeader: string,
  signingKey: Buffer,
): void {
  if (!signatureHeader.startsWith("v1=")) {
    throw authenticationFailure();
  }
  const supplied = decodeBase64Url(signatureHeader.slice(3));
  const expected = createHmac("sha256", signingKey)
    .update(`${platformContextVersion}.${payload}`)
    .digest();
  if (
    supplied.length !== expected.length
    || !timingSafeEqual(supplied, expected)
  ) {
    throw authenticationFailure();
  }
}

function verifyLifetime(context: PlatformRequestContextV1, now: Date): void {
  const issuedAt = Date.parse(context.issuedAt);
  const expiresAt = Date.parse(context.expiresAt);
  const nowMs = now.getTime();
  if (
    !Number.isFinite(issuedAt)
    || !Number.isFinite(expiresAt)
    || expiresAt <= issuedAt
    || expiresAt - issuedAt > maximumLifetimeMs
    || issuedAt > nowMs + allowedClockSkewMs
    || expiresAt <= nowMs
  ) {
    throw authenticationFailure();
  }
}

function freezeContext(
  context: z.output<typeof platformRequestContextSchema>,
): PlatformRequestContextV1 {
  return Object.freeze({
    ...context,
    actor: Object.freeze({
      ...context.actor,
      roles: Object.freeze([...context.actor.roles]),
    }),
    tenant: Object.freeze({ ...context.tenant }),
    project: Object.freeze({ ...context.project }),
    permissions: Object.freeze([...context.permissions]),
  });
}

function projectKey(request: FastifyRequest): string | undefined {
  if (typeof request.params !== "object" || request.params === null) {
    return undefined;
  }
  const value = (request.params as Record<string, unknown>).websiteProjectKey;
  return typeof value === "string" ? value : undefined;
}

export function verifyPlatformRequestContextV1(
  request: FastifyRequest,
  options: ConsumerOptions,
): PlatformRequestContextV1 {
  const signingKey = Buffer.from(options.signingKey);
  if (signingKey.length < 32) {
    throw new TypeError("Platform context signing key must contain at least 32 bytes");
  }
  const payload = headerValue(request, platformContextHeader);
  if (payload.length > 16_384) {
    throw authenticationFailure();
  }
  verifyIntegrity(
    payload,
    headerValue(request, platformContextSignatureHeader),
    signingKey,
  );

  let parsed: unknown;
  try {
    parsed = JSON.parse(decodeBase64Url(payload).toString("utf8"));
  } catch {
    throw authenticationFailure();
  }
  const result = platformRequestContextSchema.safeParse(parsed);
  if (!result.success) {
    throw authenticationFailure();
  }
  const context = freezeContext(result.data);
  verifyLifetime(context, options.now?.() ?? new Date());
  const requestedProjectKey = projectKey(request);
  if (
    requestedProjectKey !== undefined
    && requestedProjectKey !== context.project.websiteProjectKey
  ) {
    throw new BacklinkError({
      code: backlinkErrorCodes.accessDenied,
      message: "The platform context is not valid for this project.",
    });
  }
  return context;
}

function sendContextFailure(
  error: unknown,
  request: FastifyRequest,
  reply: FastifyReply,
): FastifyReply {
  const problem = toBacklinkProblemDetails(error, request.id);
  return reply
    .code(problem.status)
    .type(backlinkProblemContentType)
    .send(problem);
}

export function registerBacklinksPlatformContextConsumer(
  app: FastifyInstance,
  options: ConsumerOptions,
): void {
  if (Buffer.byteLength(options.signingKey) < 32) {
    throw new TypeError("Platform context signing key must contain at least 32 bytes");
  }
  app.decorateRequest("actor");
  app.decorateRequest("platformContext");
  app.addHook("preHandler", async (request, reply) => {
    if (projectKey(request) === undefined) {
      return;
    }
    try {
      const context = verifyPlatformRequestContextV1(request, options);
      request.platformContext = context;
      request.actor = createActorContext({
        userId: context.actor.userId,
        sessionId: context.actor.sessionId,
        roles: context.actor.roles,
      });
    } catch (error) {
      return sendContextFailure(error, request, reply);
    }
  });
}

File: backend/api/app/core/platform_request_context.py
import base64
import hashlib
import hmac
import json
import re
from collections.abc import Mapping
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta


PLATFORM_CONTEXT_HEADER = "x-growthos-platform-context"
PLATFORM_CONTEXT_SIGNATURE_HEADER = "x-growthos-platform-context-signature"
PLATFORM_CONTEXT_VERSION = "PlatformRequestContext.v1"
PLATFORM_CONTEXT_ISSUER = "growthos-platform-gateway"
PLATFORM_CONTEXT_AUDIENCE = "growthos-backlinks-core"
DEFAULT_TTL_SECONDS = 30
MAX_TTL_SECONDS = 60
PERMISSION_PATTERN = re.compile(r"^[a-z][a-z0-9_.-]*:[a-z][a-z0-9_.-]*$")


@dataclass(frozen=True)
class PlatformActor:
    user_id: str
    session_id: str
    roles: tuple[str, ...]


@dataclass(frozen=True)
class PlatformTenant:
    organization_id: str
    workspace_id: str


@dataclass(frozen=True)
class PlatformProject:
    website_project_id: str
    website_project_key: str


@dataclass(frozen=True)
class ResolvedPlatformRequestContext:
    actor: PlatformActor
    tenant: PlatformTenant
    project: PlatformProject
    permissions: tuple[str, ...]
    correlation_id: str


@dataclass(frozen=True)
class ResolvedPlatformCollectionContext:
    actor: PlatformActor
    tenant: PlatformTenant
    permissions: tuple[str, ...]
    correlation_id: str
    authorized_project_ids: tuple[str, ...] | None


def strip_untrusted_platform_context_headers(
    headers: Mapping[str, str],
) -> dict[str, str]:
    blocked = {
        PLATFORM_CONTEXT_HEADER,
        PLATFORM_CONTEXT_SIGNATURE_HEADER,
    }
    return {key: value for key, value in headers.items() if key.lower() not in blocked}


def _require_identifier(field: str, value: str) -> str:
    if not value or value != value.strip() or len(value) > 200:
        raise ValueError(f"{field} must be a non-blank identifier")
    return value


def _normalize_values(
    field: str,
    values: tuple[str, ...],
    *,
    maximum: int,
) -> list[str]:
    if not values or len(values) > maximum:
        raise ValueError(f"{field} must contain between 1 and {maximum} values")
    normalized = sorted({_require_identifier(field, value) for value in values})
    if len(normalized) != len(values):
        raise ValueError(f"{field} must not contain duplicates")
    return normalized


def _format_timestamp(value: datetime) -> str:
    if value.tzinfo is None or value.utcoffset() is None:
        raise ValueError("now must be timezone-aware")
    return value.astimezone(UTC).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def _base64url(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).rstrip(b"=").decode("ascii")


def issue_platform_request_context_v1(
    resolved: ResolvedPlatformRequestContext | ResolvedPlatformCollectionContext,
    *,
    signing_key: bytes,
    now: datetime,
    ttl_seconds: int = DEFAULT_TTL_SECONDS,
) -> dict[str, str]:
    if len(signing_key) < 32:
        raise ValueError("signing_key must contain at least 32 bytes")
    if ttl_seconds < 1 or ttl_seconds > MAX_TTL_SECONDS:
        raise ValueError(f"ttl_seconds must be between 1 and {MAX_TTL_SECONDS}")

    issued_at = _format_timestamp(now)
    expires_at = _format_timestamp(now + timedelta(seconds=ttl_seconds))
    permissions = _normalize_values("permissions", resolved.permissions, maximum=128)
    for permission in permissions:
        if not PERMISSION_PATTERN.fullmatch(permission):
            raise ValueError("permissions must use namespace:action syntax")

    payload = {
        "actor": {
            "roles": _normalize_values("actor.roles", resolved.actor.roles, maximum=32),
            "sessionId": _require_identifier("actor.sessionId", resolved.actor.session_id),
            "userId": _require_identifier("actor.userId", resolved.actor.user_id),
        },
        "audience": PLATFORM_CONTEXT_AUDIENCE,
        "correlationId": _require_identifier("correlationId", resolved.correlation_id),
        "expiresAt": expires_at,
        "issuedAt": issued_at,
        "issuer": PLATFORM_CONTEXT_ISSUER,
        "permissions": permissions,
        "project": (
            {
                "websiteProjectId": _require_identifier(
                    "project.websiteProjectId",
                    resolved.project.website_project_id,
                ),
                "websiteProjectKey": _require_identifier(
                    "project.websiteProjectKey",
                    resolved.project.website_project_key,
                ),
            }
            if isinstance(resolved, ResolvedPlatformRequestContext)
            else None
        ),
        "tenant": {
            "organizationId": _require_identifier(
                "tenant.organizationId",
                resolved.tenant.organization_id,
            ),
            "workspaceId": _require_identifier(
                "tenant.workspaceId",
                resolved.tenant.workspace_id,
            ),
        },
        "version": PLATFORM_CONTEXT_VERSION,
    }
    encoded_payload = _base64url(
        json.dumps(
            payload,
            ensure_ascii=False,
            separators=(",", ":"),
            sort_keys=True,
        ).encode("utf-8")
    )
    signature = hmac.new(
        signing_key,
        f"{PLATFORM_CONTEXT_VERSION}.{encoded_payload}".encode("ascii"),
        hashlib.sha256,
    ).digest()
    return {
        PLATFORM_CONTEXT_HEADER: encoded_payload,
        PLATFORM_CONTEXT_SIGNATURE_HEADER: f"v1={_base64url(signature)}",
    }
